vavr/src/main/java/io/vavr/collection/HashSet.java

```diff
@@ -494,11 +494,10 @@ private HashSet(HashArrayMappedTrie<T, T> tree) {
             final HashSet<T> set = (HashSet<T>) elements;
             return set;
         }
-        final HashArrayMappedTrie<T, T> that = addAll(tree, elements);
-        if (that.size() == tree.size()) {
+        if (Collections.isEmpty(elements) || this.equals(elements)){
```

Comment:
The new way is only better when this HashSet and the elements are equal but it adds an overhead for all other cases. Please restore the previous version.